backend/migrations/dataall_migrations/base_migration.py

```diff
@@ -0,0 +1,65 @@
+import logging
+import os
+
+logger = logging.getLogger()
+logger.setLevel(os.environ.get('LOG_LEVEL', 'INFO'))
+
+
+def protect(*protected):
+    """Returns a metaclass that protects all attributes given as strings"""
+
+    class Protect(type):
+        has_base = False
+
+        def __new__(meta, name, bases, attrs):
+            if meta.has_base:
+                for attribute in attrs:
+                    if attribute in protected:
+                        raise AttributeError('Overriding of attribute "%s" not allowed.' % attribute)
+            meta.has_base = True
+            klass = super().__new__(meta, name, bases, attrs)
+            return klass
+
+    return Protect
+
+
+class BaseDataAllMigration(metaclass=protect('set_next', 'set_previous', 'is_initial', 'is_last')):
+    key = '-1'
+    name = 'Base Migration'
+    description = 'Base Migration'
+
+    @classmethod
+    def up(cls):
+        logger.info(f'Upgrade is not defined for migration {cls.name}')
+
+    @classmethod
+    def down(cls):
+        logger.info(f'Downgrade is not defined for migration {cls.name}')
+
+    @classmethod
+    def set_previous(cls, previous_migration_key):
+        cls.previous_migration = previous_migration_key
+
+    @classmethod
+    def set_next(cls, next_migration_key):
+        cls.next_migration = next_migration_key
+
+    @classmethod
+    def next(cls):
+        if 'next_migration' in cls.__dict__:
+            return cls.next_migration
+        return None
+
```

Comment:
don't we need to ensure that every new migration contains the class variable `'previous_migration'`? If we return None it might break the herder and go unnoticed

Reply:
No, It won't break anything for sure. The worst case scenario the migration without parent will simply not be run.

Reply:
The point is that alembic migrations are only valid when we have a single thread of migrations: A -> B -> C -> D. If no previous migration is specified the herder won't notice it which means that we might have a broken thread of migrations: A -> B, C (without previous), C->D. In this case  the herder will just run A and B and stop there. I think we need to ensure that migrations are valid and that there is one single unbroken thread